pglock/core.py

```diff
@@ -0,0 +1,394 @@
+            lock cannot be acquired or a timeout happens.
+
+    Returns:
+        bool: When using the default side effect, returns ``True`` if the lock was acquired or
+        ``False`` if not.
+
+    Raises:
+        django.db.utils.OperationalError: When a lock cannot be acquired or a timeout happens
+            when using ``side_effect=pglock.RAISE``.
+        ValueError: If an invalid ``side_effect`` is provided or no lock ID is supplied for
+            the context manager.
+        TypeError: If the lock ID is not a string or int.
+    """  # noqa
+
+    def __init__(
+        self,
+        lock_id=None,
+        *,
+        shared=False,
+        wait=True,
+        using=DEFAULT_DB_ALIAS,
+        timeout=None,
+        side_effect=RETURN,
+    ):
+        """Acquire an advisory lock"""
+        if side_effect not in (RETURN, RAISE, SKIP):
+            raise ValueError(
+                "side_effect must be one of pglock.RETURN, pglock.RAISE, or pglock.SKIP"
+            )
+
+        self.acquire = f'pg{"_try" if not wait else ""}_advisory_lock{"_shared" if shared else ""}'
+        self.release = f'pg_advisory_unlock{"_shared" if shared else ""}'
+        self.lock_id = lock_id
+        self.using = using
+        self.timeout = timeout
+        self.side_effect = side_effect
+        self.wait = wait
+        self.shared = shared
+        self._used_as_context_manager = True
+
+    @property
+    def int_lock_id(self):
+        return _cast_lock_id(self.lock_id)
+
+    def __call__(self, func):
+        self._used_as_context_manager = False
+
+        if not self.lock_id:
+            module = inspect.getmodule(func)
+            self.lock_id = f"{module.__name__}.{func.__name__}"
+
+        @functools.wraps(func)
+        def inner(*args, **kwargs):
+            with self._recreate_cm():
+                if self.acquired or self.side_effect != SKIP:
+                    return func(*args, **kwargs)
+
+        return inner
+
+    def __enter__(self):
+        if not self.lock_id:
+            raise ValueError("Must supply a lock ID as the first argument.")
+
+        if self._used_as_context_manager and self.side_effect == SKIP:
+            raise ValueError(
+                "Cannot use pglock.SKIP in a context manager."
+                " Use the @pglock.advisory decorator instead."
+            )
+
+        sql = f"SELECT {self.acquire}({self.int_lock_id})"
+
+        with connections[self.using].cursor() as cursor:
+            try:
+                with contextlib.ExitStack() as stack:
+                    if self.timeout is not None:
+                        stack.enter_context(lock_timeout(self.timeout, using=self.using))
+
+                        if self.side_effect == RETURN and connections[self.using].in_atomic_block:
+                            # If returning True/False, create a savepoint so that
+                            # the transaction isn't in an errored state when returning.
```

Comment:
Imagine that you do this as a user:

```
with transaction.atomic():
    with pglock.advisory(timeout=timedelta(seconds=1)) as acquired:
        # Do queries here...
        User.objects...
```

By default, a lock timeout will cause the transaction to be in an error state, meaning that the `User.objects` query will fail. A save point is made around the ``pg_advisory_lock..`` statement to prevent this from happening. If a timeout occurs, code can continue as normal.

There's a test case for this in the test suite that fails

Reply:
I think there might be some confusion here. Note that I only
do the savepoint if already in a transaction. If not in a transaction, a new one won't be started.
The savepoint *only* surrounds the call to the `pg_advisory..` function too. It doesn't surround
the entire wrapped block.

Consider this:

```
with transaction.atomic():
    with pglock.advisory(timeout=1) as acquired:
        if not acquired:
            # Query some information if the advisory lock cannot be acquired
            users = User.objects.all()     
        else:
            # Do another operation if you have the lock
```

You're saying that you expect line 5 to throw an exception if the lock can't be acquired? That's what it will
do without the savepoint.

Consider this - let's make a function that grabs an advisory lock:

```
def my_function():
    with pglock.advisory(timeout=0) as acquired:
        if not acquired:
            return False
        else:
            # Do some atomic action
            return True
```

Now call that function in a transaction. If ``False`` is returned, every query will
fail afterwards.

```
@transaction.atomic()
def my_other_function():
    val = my_function()

    # Every query will fail if my_function returns False
```